packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
+  public readonly root: RootModule;
+
+  public constructor(assembly: Assembly) {
+    const packageName = assembly.name
+      .replace('@', '')
+      .replace(/[^a-z0-9_.]/gi, '');
+    const filePath = join(...packageName.split('.'));
+    const file = `${filePath}.go`;
+    super(file);
+
+    this.assembly = assembly;
+    this.root = this;
+    this.packageName = packageName;
+    // moduleName == packageName for root;
+    this.moduleName = packageName;
+    this.filePath = filePath;
+    this.file = file;
+  }
+
+  public get types(): ModuleTypes {
+    return buildModuleTypes(this, Object.values(this.assembly.types));
+  }
+
+  public get submodules(): Submodule[] {
+    return buildSubmodules(this, this, this.assembly.submodules);
+  }
+
+  public get dependencies(): Set<Module> {
+    return new Set(
+      this.types
+        .reduce(
+          (accum: Module[], t: ModuleType) => [...accum, ...t.dependencies],
+          [],
+        )
+        .filter((mod) => mod.packageName !== this.packageName),
     );
   }
 
+  public hasSubmodules(): boolean {
+    return Boolean(this.submodules.length);
+  }
+
+  public findType(fqn: string): ModuleType | undefined {
+    return findTypeInTree(this, fqn);
+  }
+
+  public emit(code: CodeMaker): void {
+    this.open(code);
+    code.line(`package ${this.packageName}`);
+    code.line();
+
+    code.line('import (');
+    code.indent();
+    code.line(`"${JSII_MODULE_NAME}"`);
+    if (this.dependencies.size > 0) {
+      this.dependencies.forEach((mod) => {
+        if (mod.packageName !== this.packageName) {
+          code.line(`"${mod.packageName}"`);
+        }
+      });
+    }
+    code.unindent();
+    code.line(')');
+    code.line();
+
+    this.emitTypes(code);
+    this.close(code);
+
+    this.emitSubmodules(code);
+  }
+
   public emitTypes(code: CodeMaker) {
     Object.values(this.types).forEach((type) => {
       type.emit(code);
     });
+  }
 
-    //     this.submodules.forEach(submodule => {
-    //       code.openFile(submodule.filename);
-    //       submodule.emitTypes(code);
-    //       code.closeFile(submodule.filename);
-    //     });
+  public emitSubmodules(code: CodeMaker) {
+    this.submodules.forEach((submodule) => {
+      submodule.emit(code);
+    });
   }
 }
 
-export class Submodule extends Module {
-  public constructor(assembly: Assembly, submodule: JsiiSubmodule) {
-    super(assembly, submodule);
+export class Submodule extends ModuleFile implements Module {
+  private readonly assembly: JsiiSubmodule;
+  public readonly packageName: string;
+  public readonly moduleName: string;
+  public readonly file: string;
+  public readonly filePath: string;
+  public readonly parent: Module;
+  public readonly root: RootModule;
+
+  public constructor(
+    root: RootModule,
+    parent: Module,
+    submodule: JsiiSubmodule,
+  ) {
+    const moduleName = submodule.name
+      .replace('@', '')
+      .replace(/[^a-z0-9]/gi, '');
```

Comment:
It's the second time I see this. Looks like this is some kind of identifier sanitization, and I'd say it should be extracted to a helper module, and unit tested.

Reply:
Wasn't this the goNameFromJSII helper method at some point? I remember it wasn't used as much as we thought (it was incorrectly being used to generate method names as well). But agree on unit testing.